mgr/cephadm: staggered upgrade

adk3798 opened the pull request:
Fixes: https://tracker.ceph.com/issues/54135

Signed-off-by: Adam King <adking@redhat.com>

~~Still needs docs and tests.~~

Additionally, an open question is how far do we want to enforce upgrade ordering? The whole list? Only up until mds? Just mgr then mon? I'm not sure, really.

For reference, upgrade order is:  mgr -> mon -> crash -> osd -> mds -> rgw -> rbd-mirror -> cephfs-mirror -> iscsi -> nfs
<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".

  - The Signed-off-by line in every git commit is important; see <span class="x x-first x-last">[Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/master/</span>SubmittingPatches.rst<span class="x x-first x-last">)</span>
-->

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [x] Updates relevant documentation
  - [ ] No doc update is appropriate
- Tests (select at least one)
  - [x] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [x] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [ ] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

vumrao commented:
YAY :+1:

adk3798 commented:
Added unit test for staggered upgrade param validation and a teuthology test for staggered upgrade.

adk3798 commented:
jenkins render docs

adk3798 commented:
jenkins render docs

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+   If you attempt to specify parameters that would force us to upgrade daemons out of
+   order, the upgrade command will block and inform you of which daemons earlier
+   in the upgrade order will be missed using your parameters.
+
```
Suggest:
>If you specify parameters that would upgrade daemons out of order, the upgrade command will block and note which will be missed if you proceed.

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+   upgrade. This may require pulling the image you are trying to upgrade to. Do
+   not be surprised if the upgrade start command takes a while to return when limiting
+   parameters are provided.
+
```
Suggest:
>Upgrade commands with
   limiting parameters will validate the options before beginning the
   upgrade, which may require pulling the new container image. Do
   not be surprised if the upgrade start command takes a while to return when limiting
   parameters are provided.

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+.. note::
+
+   In staggered upgrade scenarios (when a limiting parameter is provide) monitoring
+   stack daemons such as prometheus or node-exporter are "upgraded" (if you are
```
* `provided`
* `Prometheus`

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+   upgrading. Do not be surprised if you attempt to upgrade mgr daemons and it
+   takes longer than expected as it may just be upgrading monitoring stack daemons
+   as well.
+
```
Suggest:
>In staggered scenarios (when a limiting parameter is provided) monitoring
   stack daemons including Prometheus and node-exporter are refreshed after the Manager daemons have been upgraded. Do not be surprised if Manager upgrades thus take longer than expected. Note that the versions of monitoring stack daemons may not change between Ceph releases, in which case they are only redeployed.

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+
+While upgrading from a version that already supports staggered upgrades the process
+simply requires providing the necessary arguments. However, if you wish to upgrade
+TO a version that supports staggered upgrade from one that does not, there is a
```
* `s/TO/to`
* `s/mgr/Manager/`

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+While upgrading from a version that already supports staggered upgrades the process
+simply requires providing the necessary arguments. However, if you wish to upgrade
+TO a version that supports staggered upgrade from one that does not, there is a
+workaround that can be used. It requires first manually upgrading the mgr daemons
```
* `s/that can be used//`
* `s/mgr/Manager`

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+simply requires providing the necessary arguments. However, if you wish to upgrade
+TO a version that supports staggered upgrade from one that does not, there is a
+workaround that can be used. It requires first manually upgrading the mgr daemons
+and then passing the limiting parameters normally from that point on.
```
Suggest:
* `s/normally from that point on/as usual/`

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+To start with, determine which mgr is your active one and which are standby. This
+can be done in a variety of ways such as looking at the ``ceph -s`` output. Then,
+manually upgrade each standby mgr daemon with:
+
```
Suggest Manager instead of mgr.

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+   command may not have the ``--image`` flag. In that case, you must manually set the
+   mgr container image ``ceph config set mgr container_image <new-image-name>`` and then
+   redeploy the mgr ``ceph orch daemon redeploy mgr.example1.abcdef``
+
```
ditto for these and below instead of `mgr` that aren't part of example commands.

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+  ceph orch upgrade start --image <new-image-name> --daemon-types mgr
+
+You should now have all your mgr daemons on the new version and be able to
+make use of the limiting parameters for the rest of the upgrade.
```
`s/make use of/specify/`

anthonyeleven reviewed (changes requested):
Nice.  Upgrades and command line syntax are areas where it's really easy to be confused, so I love seeing detailed descriptions and examples like this.  Here are some nitpicky doc suggestions.

sebastian-philipp commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -261,14 +302,101 @@ def upgrade_start(self, image: str, version: str) -> str:
+                CephadmServe(self.mgr)._get_container_image_info(target_name))
+        except OrchestratorError as e:
+            raise OrchestratorError(f'Failed to pull {target_name}: {str(e)}')
+        # what we need to do here is build a list of daemons that must already be upgraded
+        # in order for the user's selection of daemons to upgrade to be valid. for example,
+        # if they say --daemon-types 'osd,mds' but mons have not been upgraded, we block.
```
:heart:

adk3798 commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+Some users may prefer to have their upgrade be done in shorter bursts rather
+than one fire-and-forget upgrade command. The upgrade command, starting
+in 16.2.9 and 17.2.1 allows specifying parameters to limit which daemons are
+upgraded by a single upgrade command. The options in include ``daemon_types``,
```
The cephadm upgrade does only upgrade daemons. In this situation the hosts parameter is specifying to only upgrade daemons on that given host, not the hosts themselves.

adk3798 commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,101 @@ you need. For example, the following command upgrades to a development build:
+at a time (e.g. can't provide an osd and rgw service at the same time) and
+will only upgrade daemons belonging to those services. ``hosts`` can be combined
+with ``daemon_types`` or ``services`` or provided on its own. The argument
+follows the same format as the command line options for :ref:`orchestrator-cli-placement-spec`.
```
this was intended to just refer to just the single `hosts` parameter. I updated it so maybe it's more clear now.

adk3798 commented:
@anthonyeleven I think I've added in your recommended changes (and left comments on the ones I didn't) and also added documentation for the new `--limit` parameter that has since been added here. Can you give this another look?

Also, thanks for reviewing this in general. Appreciate someone with docs specific expertise looking over this.

adk3798 commented:
new staggered upgrade test passing (as well as standard upgade case)

https://pulpito.ceph.com/adking-2022-05-02_20:38:27-orch:cephadm:upgrade-wip-adk2-testing-2022-05-02-1133-distro-basic-smithi/

adk3798 commented:
https://pulpito.ceph.com/adking-2022-05-04_19:11:28-orch:cephadm-wip-adk-testing-2022-05-04-1135-distro-basic-smithi/

green

rkachach commented on qa/suites/orch/cephadm/upgrade/3-upgrade/staggered.yaml:
```diff
@@ -0,0 +1,78 @@
+      - ceph versions | jq -e '.mgr | length == 1'
+      - ceph versions | jq -e '.mgr | keys' | grep $sha1
+      - ceph versions | jq -e '.overall | length == 2'
+      - ceph orch upgrade check quay.ceph.io/ceph-ci/ceph:$sha1 | jq -e '.up_to_date | length == 2'
```
For all the similar checks it would be good to indicate what are we checking for (why are we expecting a `length == 2`).

adk3798 replied:
added a bunch of comments to this test to explain what each part is checking

rkachach commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -272,14 +322,103 @@ def upgrade_start(self, image: str, version: str) -> str:
+            for daemon_type in CEPH_UPGRADE_ORDER[::-1]:
+                if daemon_type in dtypes:
+                    return daemon_type
+            return 'N/A'
```
is there any special reason for not using `None` in this case instead of `N/A`?

adk3798 replied:
no real reason for `'N/A'` in particular. Wanted it to not be `None` so I didn't have to worry about typing and would always get a string back, but honestly it should really just be an empty string by that logic rather than `'N/A'`. I'll make that change,

adk3798 replied:
changed this to return an empty string instead

rkachach commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -272,14 +322,103 @@ def upgrade_start(self, image: str, version: str) -> str:
+        dtypes = []
+        if daemon_types is not None:
+            dtypes = daemon_types
+            if hosts is not None:
+                dtypes = [_latest_type(dtypes)]
+                other_host_daemons = [
+                    d for d in daemons if d.hostname is not None and d.hostname not in hosts]
+                daemons = _get_earlier_daemons(dtypes, other_host_daemons)
+            else:
+                daemons = _get_earlier_daemons(dtypes, daemons)
```
This code portion seems to be mostly the same in the three different cases. Maybe it could be factorized out from the if clauses. Se we only keep the `dtypes` calculation which seems to be different for each case.

adk3798 replied:
I tried looking through the 3 cases to see which subsections could maybe be their own function. The only thing I could ifnd really repeated was
```
                other_host_daemons = [
                    d for d in daemons if d.hostname is not None and d.hostname not in hosts]
                daemons = _get_earlier_daemons(dtypes, other_host_daemons)
            else:
                daemons = _get_earlier_daemons(dtypes, daemons)
```
but even then it's a bit annoying to try to split that section off into its own function as dtypes is being set differently in each case (within the if block in one case, outside in  the other) and that behavior would have to be preserved. What did you have in mind here in terms of splitting things up?

rkachach replied:
I was not referring to get each section out in a separate function but re-arranging the code to make it clear that what really changes from each if clause is the `dtypes` calculation. I worked out a little big the logic and got the following, Ithink it's equivalent. BTW: I don't understand why in `elif services is not None` we don't use the `_latest_type(dtypes)` meanwhile in the other cases we do use it. Just a comment to make sure we are not forgetting to apply it for that case.

```

        if daemon_types is not None:
            dtypes = daemon_types if hosts is not None else [_latest_type(dtypes)]
            err_msg_base += 'Daemons with types earlier in upgrade order than given types need upgrading.\n'
        elif services is not None:
            # for our purposes here we can effectively convert our list of services into the
            # set of daemon types the services contain. This works because we don't allow --services
            # and --daemon-types at the same time and we only allow services of the same type
            sspecs = [self.mgr.spec_store[s].spec for s in services if self.mgr.spec_store[s].spec is not None]
            stypes = list(set([s.service_type for s in sspecs]))
            if len(stypes) != 1:
                raise OrchestratorError('Doing upgrade by service only support services of one type at a time. Found service types: {stypes}')
            for stype in stypes:
                dtypes += orchestrator.service_to_daemon_types(stype)
            dtypes = list(set(dtypes))
            err_msg_base += 'Daemons with types earlier in upgrade order than daemons from given services need upgrading.\n'
        elif hosts is not None:
            # hosts must be handled a bit differently. For this, we really need to find all the daemon types
            # that reside on hosts in the list of hosts we will upgrade. Then take the type from
            # that list that is latest in the upgrade order and check if any daemons on hosts not in the
            # provided list of hosts have a daemon with a type earlier in the upgrade order that is not upgraded.
            dtypes = list(set([d.daemon_type for d in daemons if d.daemon_type is not None and d.hostname in hosts]))
            dtypes = _latest_type(dtypes)
            err_msg_base += 'Daemons with types earlier in upgrade order than daemons on given host need upgrading.\n'

        other_host_daemons = daemons
        if hosts is not None:
            other_host_daemons = [d for d in daemons if d.hostname is not None and d.hostname not in hosts]
        daemons = _get_earlier_daemons(dtypes, other_host_daemons)

```

adk3798 replied:
it's unnecessary to do that for services because we only allow services of one daemon type at a time with that flag. The reason being allowing services of different types makes the verification significantly more complex.

rkachach replied:
got it, thanks for the clarification :+1:

rkachach commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -534,6 +673,273 @@ def _enough_mds_for_ok_to_stop(self, mds_daemon: DaemonDescription) -> bool:
+        if target_digests is None:
+            target_digests = []
+        for d_entry in to_upgrade:
+            if self.upgrade_state.remaining_count is not None and self.upgrade_state.remaining_count <= 0 and not d_entry[1]:
```
what's `d_entry[1] (and d_entry[0])` could we use some named tuple (or a class) instead?

adk3798 replied:
we could, but these tuples aren't new in this change, I just moved the functionality into a separate function so, while we could definitely refactor it a bit, I'd rather not do it in this already large PR.

rkachach commented on qa/suites/orch/cephadm/upgrade/3-upgrade/staggered.yaml:
```diff
@@ -0,0 +1,78 @@
+tasks:
```
As I understood from reading the docs we are adding different types of staggered upgrade:

- by daemon_types
- by services
- by hosts
- by limit

Instead of having all of them in the same (big) test case I think it would be better to have different (smaller) test cases for each one of them. That would make it easier to debug issues (in case of future failures) and would make it easier for anybody trying to read/understand the test case.

adk3798 replied:
the reason I wanted one big test here is because I didn't want to have a bunch of different tests. Teuthology tests require a decent more actual lab resources than unit tests so it's preferable to have one test for the new feature rather than  adding 4 different tests imo.

rkachach replied:
I see, I'm afraid it would be hard to debug failures on this test-case being so long. Anyway in the worst case we can modify it during the debugging to keep only the failing part :+1:

adk3798 commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -534,6 +673,273 @@ def _enough_mds_for_ok_to_stop(self, mds_daemon: DaemonDescription) -> bool:
 
         return True  # if mds has no fs it should pass ok-to-stop
 
+    def _detect_need_upgrade(self, daemons: List[DaemonDescription], target_digests: Optional[List[str]] = None) -> Tuple[bool, List[Tuple[DaemonDescription, bool]], List[Tuple[DaemonDescription, bool]], int]:
```
added some code comments at the top of the function to explain what this does

adk3798 commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -272,14 +322,103 @@ def upgrade_start(self, image: str, version: str) -> str:
+            latest = _latest_type(dtypes)
+            if latest == 'N/A':
+                return []
+            earlier_types = '|'.join(CEPH_UPGRADE_ORDER).split(latest)[0].split('|')[:-1]
```
added a code comments to the function to explain its purpose

rkachach replied:
I see, I agree on addressing the refactoring as a separate PR

adk3798 commented on src/pybind/mgr/cephadm/upgrade.py:
```diff
@@ -146,7 +180,18 @@ def _get_upgrade_info(self) -> Tuple[str, List[str]]:
+            daemons = [d for d in self.mgr.cache.get_daemons(
+            ) if d.daemon_type in CEPH_UPGRADE_ORDER]
+        if self.upgrade_state.hosts is not None:
+            daemons = [d for d in daemons if d.hostname in self.upgrade_state.hosts]
```
moved this to its own function

adk3798 commented:
@rkachach I think I've responded to or addressed all of your review comments. Let me know if there's anything else or questions on what I said,

adk3798 commented:
jenkins test dashboard cephadm

adk3798 commented:
jenkins test dashboard cephadm

adk3798 commented:
ping @anthonyeleven I think this PR is getting close to being fully ready if you have a chance to re-review the docs

anthonyeleven commented on doc/cephadm/upgrade.rst:
```diff
@@ -188,3 +188,100 @@ you need. For example, the following command upgrades to a development build:
+Some users may prefer to upgrade components in phases rather than all at once.
+The upgrade command, starting in 16.2.9 and 17.2.1 allows specifying parameters
+to limit which daemons areupgraded by a single upgrade command. The options in
+include ``daemon_types``, ``services``, ``hosts`` and ``limit``. ``daemon_types``
```
space `are upgraded`, maybe comma after 17.2.1
Maybe even take out `specifying`

adk3798 replied:
done

anthonyeleven commented:
LGTM for docs, with the above minor suggestions.

adk3798 commented:
jenkins test dashboard cephadm

adk3798 commented:
20 runs of new test for staggered upgrade, all passed

https://pulpito.ceph.com/adking-2022-05-20_04:04:42-orch:cephadm:upgrade-wip-adk-testing-2022-05-19-2005-distro-basic-smithi/
https://pulpito.ceph.com/adking-2022-05-20_05:14:33-orch:cephadm:upgrade-wip-adk-testing-2022-05-19-2005-distro-basic-smithi/

Additional full orch run, all passed

https://pulpito.ceph.com/adking-2022-05-20_15:32:09-orch:cephadm-wip-adk-testing-2022-05-19-2005-distro-basic-smithi/
